### PR TITLE
Update to Jest 29

### PR DESCRIPTION
**Pull request by ymqy**
## Summary

- yarn.lock diff +-6249, **small pr**
- use jest-environment-jsdom by default
- uncaught error from jsdom is an error object instead of strings
- abortSignal.reason is read-only in jsdom and node, https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/reason

## How did you test this change?

ci green

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js
Comment (ymqy) [inline packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js:20]: https://github.com/facebook/react/pull/21575#discussion_r755474351

File: packages/scheduler/src/__tests__/SchedulerSetTimeout-test.js
Comment (ymqy) [inline packages/scheduler/src/__tests__/SchedulerSetTimeout-test.js:26]: Jest v28 add MessageChannel, MessageEvent to globals. https://github.com/facebook/jest/pull/12553

File: packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js
Comment (ymqy) [inline packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js:8]: In Jest v26, the module path 'react-dom/server' would map to the file path 'packages/react-dom/server.js'. However, in Jest v29, this mapping rule has changed and it will first look at the exports field of the package.json file, and then decide which file to load based on the testing environment. In jest-environment-jsdom, 'packages/react-dom/server.browser.js' will be loaded, and in jest-environment-node, 'packages/react-dom/server.node.js' will be loaded. This will cause tests that use node environment methods (such as renderToPipeableStream) to fail. Is it possible to use the module path 'react-dom/server.node'? I see that the problem can also be solved by customizing the testing environment (https://github.com/facebook/react/pull/21575), adding some jsdom environment variables in the node testing environment.

File: packages/react-dom/src/__tests__/ReactDOMFizzStaticNode-test.js
Comment (ymqy) [inline packages/react-dom/src/__tests__/ReactDOMFizzStaticNode-test.js:24]: AbortController in Node environments has inconsistent behavior, for example in versions v.15.14.0 ~ v16.13.1, AbortSignal.abort() does not support the "reason" parameter, but in version v16.16.0 and above it has been tested to support the "reason" parameter. Therefore, using "abortcontroller-polyfill" ensures consistent behavior across all Node versions, helping to maintain compatibility.

File: packages/react-dom/src/__tests__/ReactDOMFizzServerBrowser-test.js
Comment (ymqy) [inline packages/react-dom/src/__tests__/ReactDOMFizzServerBrowser-test.js:234]: abortSignal.reason is read-only in jsdom and node, https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/reason

File: packages/react-dom/src/__tests__/ReactDOMServerIntegrationAttributes-test.js
Comment (ymqy) [inline packages/react-dom/src/__tests__/ReactDOMServerIntegrationAttributes-test.js:22]: "jest.resetModuleRegistry" has been removed in Jest v27 and should be replaced with "jest.resetModules".

File: packages/react-dom/src/__tests__/ReactDOMServerPartialHydration-test.internal.js
Comment (ymqy) [inline packages/react-dom/src/__tests__/ReactDOMServerPartialHydration-test.internal.js:3123]: Use the Event from jsdom instead of the one from Node.

File: packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js
Comment (ymqy) [inline packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: The following are sources of the OOM problem. Interestingly, using 'yarn test' does not cause an OOM problem, but using the following commands does: 'yarn test -r=experimental --env=development --persistent --ci', 'yarn test -r=experimental --env=production --ci', 'yarn test --build -r=experimental --env=development --ci', 'yarn test --build -r=experimental --env=production --ci'. It is speculated that this may be because the OOM problem is avoided due to mocking of certain modules in the 'scripts/jest/setupTests.[www.js](http://www.js/)' file that is referenced in 'yarn test'.

**Comment by ymqy** [comment]
@gaearon  This PR is important because Jest v29 has been released and the new version contains a lot of new features and bug fix which is important for our project. It's ready for review and merge.

**Review by sebmarkbage — APPROVED** [review]

**Review by eps1lon — COMMENTED** [review]
Can we do fixtures/ in a follow-up? These are not automatically tested and rarely used so I'd rather do some manual testing so we don't accidentally break them. 

And that we can use the automated tests in packages/ as a canary.

File: packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js
Comment (eps1lon) [inline packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js:54]:
```suggestion
    JSDOM = require('jsdom').JSDOM;
```

File: packages/react-dom/src/__tests__/ReactDOMFizzStaticNode-test.js
Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMFizzStaticNode-test.js:24]: Can we somehow hook into the Jest's standard node environment and make sure we always do this? Seems like this is something we'll run into frequently otherwise.

File: packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js
Comment (eps1lon) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: Why are we only using `removeNonEnumerableProperties` on some assertions? We have >100 assertions on `ReactNoop.getChildren()` but only a couple of are now wrapped in `removeNonEnumerableProperties`.

File: packages/react-dom/src/__tests__/ReactDOMFizzStaticNode-test.js
Comment (ymqy) [reply packages/react-dom/src/__tests__/ReactDOMFizzStaticNode-test.js:24]: image

Okay, I'll add the abortcontroller-polyfill in the scripts/jest/setupEnvironment.js file, so we can make sure it's always effective in Jest's standard node environment.

File: packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js
Comment (ymqy) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: Thanks for pointing out the issue with the assertions. 

I did some further investigation and found that the OOM problem was caused by the tests with the titles "keeps showing an avoided parent fallback if it is already showing" and "do not show placeholder when mounting an avoided boundary with startTransition".

I located the specific lines of code that were causing the problem and initially wrapped them with removeNonEnumerableProperties, but upon further review, I realized that this may not be the best solution. 

I think it would be better to further investigate and find the root cause of the problem.

Comment (eps1lon) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: Did you get OOM locally or in CI?

Comment (ymqy) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: You can reproduce it locally with the ci test command

https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621083
https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621073
https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621072
https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621071
https://app.circleci.com/pipelines/github/facebook/react/37815/workflows/af2d09e1-48d7-4e29-a33a-f4106c5783cc/jobs/621053

Comment (eps1lon) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: Not getting an OOM locally but without `removeEnumerableProperties` it takes 3 minutes for this test (to fail as expected).

Checking out what is wrong. Might be worth to convert all `expect(ReactNoop.getChildren()).toEqual` to `expect(ReactNoop).toMatchRenderedOutput` first. But we also need to understand what perf bottleneck we're hitting here.

Comment (eps1lon) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: Seems like some perf regression in `expect` when `toEqual` does not pass. 

The "received" object contains fiber nodes which are nodes in a double linked list. And it seems like `jest-matcher-utils` is calling `deepCyclicCopyObject` (from `printDiffOrStringify`) in `node_modules/jest-matcher-utils/build/deepCyclicCopyReplaceable.js` thousands of times.

I'll create branch to highlight this for Jest people. 

Failing tests are intentionally part of our test suite. But if Jest 29 introduces such severe perf regressions (going from 50ms tests to 150000ms tests) for failing `toEqual` assertions, we can't merge it in this form.

Thank you very much for highlighting this issue!

Comment (eps1lon) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: @SimenB 
```
$ git clone https://github.com/eps1lon/react.git
$ git checkout jestbump-perf-repro
$ yarn
$ yarn test -r=experimental --env=development --persistent --ci packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js --verbose
```

I'm also getting a lot of deprecation warnings from `testcheck` which is used by Jasmine. So maybe it's actually a Jasmine issue?

Comment (ymqy) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: I couldn't agree more with your concern about the performance degradation of the toEqual assertion, and will consider merging this pr after jest fixes the issue, very interested in following up on the jest performance issue fix process

Comment (ymqy) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: Before is also the best efforts to analyze the results of almost, but afraid to say not clear did not put forward. https://github.com/facebook/react/pull/26041#issuecomment-1406904681

Comment (eps1lon) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: We probably have this problem on `main` (i.e. Jest 26) already: https://app.circleci.com/pipelines/github/facebook/react/38061/workflows/a2043dd0-df62-49d5-a10e-9b534c33cbdb/jobs/623788/parallel-runs/5?filterBy=ALL

Going to convert `expect(ReactNoop.getChildren()).toEqual` to `expect(ReactNoop).toMatchRenderedOutput` in a separate PR. This will hopefully fix issues on `main` and unblock this PR.

Comment (eps1lon) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js:2431]: Settled on using `toMatchRenderedOutput` instead.

Comment (SimenB) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js]: Is there a smaller reproduction than react?

**Review by eps1lon — APPROVED** [review]
Works like a charm. Much appreciated, thanks!

File: packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js
Comment (eps1lon) [reply packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js]: I don't have one and I expect it to be quite hard to reproduce. From what I could tell, the doubly linked-list structure of Fiber caused the severe slowdown in `expect(somethingHavingFibers).toEqual()`.

**Comment by SimenB** [comment]
🎉

**Comment by kassens** [comment]
@ymqy thanks for the huge update and debugging the issues!

**Comment by gaearon** [comment]
amazing work. thank you.

**Pull request merged**
